### ISSUE TITLE
X509 certificate module broken

### ISSUE THREAD

**Issue by 1nfiniteloop** [D-Programming-Deimos/openssl#49]
**Summary**
Some of the `X509`-functions cause crashes. It is not possible to generate valid X509 certificates.

**Analysis**
The X509 struct has incorrect definitions where some members are faulty declared as pointers instead of struct members. Compare struct definition `x509_st` from C API header @ <https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/include/internal/x509_int.h#L158>
with `x509_st` in D module @ <https://github.com/D-Programming-Deimos/openssl/blob/master/deimos/openssl/x509.d#L284>

**Suggested solution**
Fix struct by removing pointer definition (*) accordingly.

**How to reproduce**
```dlang
#!/usr/bin/env dub
/+ dub.sdl:
	name "deimos.openssl-bug"
    dependency "openssl" version="~>2.0.0"
+/

import deimos.openssl.ssl;


void main()
{
    X509* certificate = X509_new();
    // Below causes a crash since pointer "cert_info" is null, it shall instead be a struct member. 
    // X509_get_notBefore --> auto X509_get_notBefore()(x509_st* x) { return (x.cert_info.validity.notBefore); }
    const ASN1_TIME *time = X509_get_notBefore(certificate);
    X509_free(certificate);
}
```

**Comment by jeking3** [comment]
In openssl 1.0.1 these structures were public.
In openssl 1.0.2 and later these structures were made private and a new set of X509 APIs added to manipulate them.

The headers you noted are for 1.1.1 and not 1.1.0h (which master is based on), and internal headers to openssl we should not depend on access to.

These structure definitions should be removed from the D implementation and left opaque.  The D implementation should be be allowing direct modification of these structures when supporting openssl >= 1.0.2 (like the current master).  There is likely more to do than just these two, but that's the general direction that should happen to make it compatible with the openssl >= 1.0.2 API.

**Comment by 1nfiniteloop** [comment]
Fixed in #54,  closing issue.
